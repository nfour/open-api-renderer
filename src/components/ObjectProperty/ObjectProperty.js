import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class ObjectProperty extends PureComponent {
  render() {
    const { constraints } = this.props;

    if (!constraints) {
      return null;
    }

    const { minProperties, maxProperties } = constraints;
    const validations = [];

    if (maxProperties !== undefined && minProperties !== undefined) {
      // Be succint if the minProperties is the same maxProperties
      // ie. value can only be of `x` length.
      if (maxProperties === minProperties) {
        validations.push(`${minProperties} properties`);
      } else {
        validations.push(`${minProperties}-${maxProperties} properties`);
      }
    } else if (minProperties !== undefined) {
      validations.push(`at least ${minProperties} properties`);
    } else if (maxProperties !== undefined) {
      validations.push(`at most ${maxProperties} properties`);
    }

    if (!validations.length) {
      return null;
    }

    return (
      <span>
        {validations.map(constraint =>
          <span key={constraint} className="object-constraints">{constraint}</span>
        )}
      </span>
    );
  }
}

ObjectProperty.propTypes = {
  constraints: PropTypes.shape({
    maxProperties: PropTypes.number,
    minProperties: PropTypes.number
  })
};
