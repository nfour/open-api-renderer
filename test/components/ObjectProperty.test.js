import React from 'react';
import ObjectProperty from './../../src/components/ObjectProperty/ObjectProperty';
import renderer from 'react-test-renderer';

describe('<ObjectProperty />', () => {
  it('renders nothing if there are no constraints', () => {
    const tree = renderer.create(
      <ObjectProperty />
    ).toJSON();

    expect(tree).toBeNull();
  });

  it('renders minProperties', () => {
    const tree = renderer.create(
      <ObjectProperty constraints={{minProperties: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders maxProperties', () => {
    const tree = renderer.create(
      <ObjectProperty constraints={{maxProperties: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minProperties and maxProperties', () => {
    const tree = renderer.create(
      <ObjectProperty constraints={{minProperties: 4, maxProperties: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minProperties and maxProperties when they are the same value', () => {
    const tree = renderer.create(
      <ObjectProperty constraints={{minProperties: 4, maxProperties: 4}} />
    );

    expect(tree).toMatchSnapshot();
  });
});
