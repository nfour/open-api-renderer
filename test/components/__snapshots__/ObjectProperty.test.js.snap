// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ObjectProperty /> renders maxProperties 1`] = `
<span>
  <span
    className="object-constraints"
  >
    at most 10 properties
  </span>
</span>
`;

exports[`<ObjectProperty /> renders minProperties 1`] = `
<span>
  <span
    className="object-constraints"
  >
    at least 2 properties
  </span>
</span>
`;

exports[`<ObjectProperty /> renders minProperties and maxProperties 1`] = `
<span>
  <span
    className="object-constraints"
  >
    4-10 properties
  </span>
</span>
`;

exports[`<ObjectProperty /> renders minProperties and maxProperties when they are the same value 1`] = `
<span>
  <span
    className="object-constraints"
  >
    4 properties
  </span>
</span>
`;
